wolfcrypt/src/port/caam/caam_driver.c

```diff
@@ -438,95 +438,93 @@ static void print_jdkek()
     CAAM_OP | CAAM_CLASS1 | CAAM_RNG | 0x00001000   /* Generate secure keys */
 };
 
 
-
 /* Initialize CAAM RNG
  * returns 0 on success */
 int caamInitRng(struct CAAM_DEVICE* dev);
 int caamInitRng(struct CAAM_DEVICE* dev)
 {
     DESCSTRUCT desc;
-    unsigned int reg, status;
-    int ret = 0;
+    unsigned int reg, entropy_delay;
+    int ret = 0, i;
 
+    /* set up the job description for RNG initialization */
     memset(&desc, 0, sizeof(DESCSTRUCT));
+    desc.desc[desc.idx++] = CAAM_HEAD; /* later will put size to header*/
+    for (i = 1; i < 6; i = i + 1) {
+        desc.desc[desc.idx++] = wc_rng_start[i];
+    }
+    desc.caam = dev;
+
+    /* Attempt to start the RNG, first trying the fastest entropy delay value
+     * and increasing it after each failed attempt until either a success is hit
+     * or the max delay value is.
+     */
+    for (entropy_delay = CAAM_ENT_DLY; entropy_delay <= CAAM_ENT_DLY_MAX;
+            entropy_delay = entropy_delay + CAAM_ENT_DLY_INCREMENT) {
+
+        /* Set up use of the TRNG for seeding wolfSSL HASH-DRBG */
+        /* check out the status and see if already setup */
+        CAAM_WRITE(CAAM_RTMCTL, CAAM_PRGM);
+        CAAM_WRITE(CAAM_RTMCTL, CAAM_READ(CAAM_RTMCTL) | 0x40); /* reset */
+
+        /* Set up reading from TRNG */
+        CAAM_WRITE(CAAM_RTMCTL, CAAM_READ(CAAM_RTMCTL) | CAAM_TRNG);
+
+        /* Set up delay for TRNG
+         * Shift left with RTSDCTL because 0-15 is for sample number
+         * Also setting the max and min frequencies */
+        CAAM_WRITE(CAAM_RTSDCTL, (entropy_delay << 16) | CAAM_ENT_SAMPLE);
+        CAAM_WRITE(CAAM_RTFRQMIN, entropy_delay >> CAAM_ENT_MINSHIFT);
+        CAAM_WRITE(CAAM_RTFRQMAX, entropy_delay << CAAM_ENT_MAXSHIFT);
+
+    #ifdef WOLFSSL_CAAM_PRINT
+        printf("Attempt with entropy delay set to %d\n", entropy_delay);
+        printf("Min delay of %d and max of %d\n",
+                entropy_delay >> CAAM_ENT_MINSHIFT,
+                entropy_delay << CAAM_ENT_MAXSHIFT);
+    #endif
 
-    /* Set up use of the TRNG for seeding wolfSSL HASH-DRBG */
-    /* check out the status and see if already setup */
-    CAAM_WRITE(CAAM_RTMCTL, CAAM_PRGM);
-    CAAM_WRITE(CAAM_RTMCTL, CAAM_READ(CAAM_RTMCTL) | 0x40); /* reset */
-
-    /* Set up reading from TRNG */
-    CAAM_WRITE(CAAM_RTMCTL, CAAM_READ(CAAM_RTMCTL) | CAAM_TRNG);
-
-    /* Set up delay for TRNG @TODO Optimizations?
-     * Shift left with RTSDCTL because 0-15 is for sample number
-     * Also setting the max and min frequencies */
-    CAAM_WRITE(CAAM_RTSDCTL, (CAAM_ENT_DLY << 16) | 0x09C4);
-    CAAM_WRITE(CAAM_RTFRQMIN, CAAM_ENT_DLY >> 1); /* 1/2      */
-    CAAM_WRITE(CAAM_RTFRQMAX, CAAM_ENT_DLY << 3); /* up to 8x */
-
-    /* Set back to run mode and clear RTMCL error bit */
-    reg = CAAM_READ(CAAM_RTMCTL) ^ CAAM_PRGM;
+        /* Set back to run mode and clear RTMCL error bit */
+        reg = CAAM_READ(CAAM_RTMCTL) ^ CAAM_PRGM;
```

Comment:
'^' => '|'

Reply:
maybe '&' ~CAAM_PRGM, was using it to clear the program bit set